plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -196,49 +195,81 @@ private CollectedStatistics mergeStatisticsIfNecessary(
-                                    .collect(toImmutableList());
-                            for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
-                                Integer fieldId = getOnlyElement(read.first().inputFields());
-                                checkState(pendingPreviousNdvSketches.remove(fieldId), "Unwanted read of stats for field %s", fieldId);
-                                Memory memory = Memory.wrap(ByteBuffers.getBytes(read.second())); // Memory.wrap(ByteBuffer) results in a different deserialized state
-                                CompactSketch previousSketch = CompactSketch.wrap(memory);
-                                CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
-                                ndvSketches.put(fieldId, SetOperation.builder().buildUnion().union(previousSketch, newSketch));
+            if (hasUsefulData) {
+                try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
+                        .withFileSize(statisticsFile.fileSizeInBytes())
+                        .withFooterSize(statisticsFile.fileFooterSizeInBytes())
+                        .build()) {
+                    List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
+                            .collect(toImmutableList());
+                    for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
+                        Integer fieldId = getOnlyElement(read.first().inputFields());
+                        if (pendingPreviousNdvSketches.remove(fieldId)) {
+                            CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
+                            if (shouldMergeWithOld) {
+                                newSketch = SetOperation.builder().buildUnion().union(readPreviousSketch(read), newSketch);
                             }
+                            ndvSketches.put(fieldId, newSketch);
                         }
-                        catch (IOException exception) {
-                            throw new UncheckedIOException(exception);
+                        else {
+                            ndvSketches.put(fieldId, readPreviousSketch(read));
                         }
                     }
                 }
-
-                yield new CollectedStatistics(ndvSketches.buildOrThrow());
+                catch (IOException exception) {
+                    throw new UncheckedIOException(exception);
+                }
+                pendingPreviousNdvSketches.forEach(newFieldId -> ndvSketches.put(newFieldId, collectedNdvSketches.get(newFieldId)));
```

Comment:
this is the case where we didn't find the other stat for merging

does not look correct when `shouldMergeWithOld` is true

Reply:
I think this problem comes from bad naming. 
We should always take recently computed stats, especially when there was no previous version